miriway.cpp

```diff
@@ -252,13 +301,25 @@ int main(int argc, char const* argv[])
 
     // A configuration option to start applications when compositor starts and record them in `shell_pids`.
     // Because of the previous section, this allows them some extra Wayland extensions
-    ShellComponents shell_components{[&](auto cmd){ shell_pids.insert(client_launcher.launch(cmd)); }};
-    runner.add_start_callback([&]{ shell_components.launch_all(); });
+    ShellComponents shell_components{[&](auto cmd){
+        pid_t pid = client_launcher.launch(cmd);
+        shell_pids.insert(pid);
+        return pid;
+    }};
+    runner.add_start_callback([&]{ shell_components.run(); });
+    runner.add_stop_callback([&] { shell_components.stop(); });
     ConfigurationOption components_option{
         [&](std::vector<std::string> const& apps) { shell_components.populate(apps); },
         "shell-component",
         "Shell component to launch on startup (may be specified multiple times)"};
 
+    ConfigurationOption auto_restart_components_option{
+        [&](std::vector<std::string> const& apps) {
+            shell_components.populate_auto_restart(apps);
+        },
+        "shell-component-auto-restart",
+        "Shell component to launch on startup that will automatically restart if it crashes (may be specified multiple times)"};
```

Comment:
I prefer what the above says but the code implements "automatically restart if it `exit`s (with or without success)".